src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py

```diff
@@ -0,0 +1,469 @@
+
+        adapter_name = powershell(["Get-NetAdapter", "-Name", name_wildcard])
+        if adapter_name == "":
+            powershell(["Add-VMNetworkAdapter",
+                        "-SwitchName", self.mgmt_vswitch_name,
+                        "-Name", self.MGMT_HOST_ADAPTER_NAME,
+                        "-ManagementOS"])
+
+        current_ip = powershell(["Get-NetAdapter -Name {} | "
+                                 "Get-NetIPAddress | "
+                                 "Where AddressFamily -eq IPv4 | "
+                                 "Select -ExpandProperty IPAddress"
+                                 .format(name_wildcard)])
+        if current_ip != self.HOST_MGMT_IP:
+            if_index = powershell(["Get-NetAdapter -Name {} | "
+                                   "Select -ExpandProperty ifIndex"
+                                   .format(name_wildcard)])
+            if current_ip != "":
+                powershell(["Get-NetAdapter -Name {} | "
+                            "Remove-NetIPAddress -Confirm:$false"
+                            .format(name_wildcard)])
+            powershell(["New-NetIPAddress", "-IPAddress", self.HOST_MGMT_IP,
+                        "-PrefixLength", str(self.MGMT_PREFIX_LEN),
+                        "-InterfaceIndex", if_index])
+
+
+    def _generate_new_mgmt_ip(self):
+        """generates an IP to give to new gateway VM"""
+        used_ips = self._get_all_mgmt_ips()
+
+        used_ips = netaddr.IPSet(used_ips)
+        pool_ips = netaddr.IPSet(self.MGMT_IP_RANGE)
+        free_ips = pool_ips ^ used_ips
+
+        try:
+            first_available_ip = next(ip for ip in free_ips)
+        except StopIteration:
+            raise ValueError("Ran out of management IPs for gateway VM")
+        return first_available_ip
+
+
+    def _get_mgmt_ip(self):
+        """queries hyper-v for management IP of SNAT VM"""
+        return self._get_mgmt_ips_of(self.wingw_name)
+
+
+    def _get_all_mgmt_ips(self):
+        """queries hyper-v for management IPs of all SNAT VMs"""
+        try:
+            mgmt_ips = self._get_mgmt_ips_of(self.WINGW_PREFIX + "*")
+        except IndexError:
+            # ignore, just return an empty list
+            mgmt_ips = []
+        return mgmt_ips
+
+
+    def _get_mgmt_ips_of(self, vmname_with_wildcard):
+        ips = powershell(["Get-VMNetworkAdapter -VMName {} | "
+                          "Where SwitchName -eq {} | "
+                          "Select -ExpandProperty IPAddresses"
+                          .format(vmname_with_wildcard, 
+                                  self.mgmt_vswitch_name)])
+        if ips == "":
+            raise IndexError("no management IP found")
+        ips = ips.splitlines()
+
+        # if multiple IPs connected to mgmt switch, we can use either one
+        return ips[0]
+
+
+    def _configure_vm_mgmt_ip(self):
+        time.sleep(self.WAIT_FOR_VM_TIME_SEC)
```

Comment:
meh :/ can we query hyper-v somehow? or maybe at the very least do ping polling?